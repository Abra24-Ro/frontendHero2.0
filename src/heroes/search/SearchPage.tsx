import { useQuery } from "@tanstack/react-query";
import { CustomBreadcrumb } from "../../components/custom/CustomBreadcrumb";
import { CustomJumbotron } from "../../components/custom/CustomJumbotron";
import { HeroStats } from "../components/HeroStats";
import SearchControls from "./ui/SearchControls";
import { useSearchParams } from "react-router";
import { searchHeroesAction } from "../actions/search.heros.action";
import { HeroGrid } from "../components/HeroGrid";

const SearchPage = () => {
  const [searchParams] = useSearchParams();

  const name = searchParams.get("name") || "";
  const strength = searchParams.get("strength") || "";

  const { data: heroes = [] } = useQuery({
    queryKey: ["search", { name, strength }],
    queryFn: () => searchHeroesAction({ name, strength }),
    staleTime: 1000 * 60 * 5,
  });
  return (
    <>
      <CustomJumbotron
        titulo="Universo de Héroes y Villanos"
        descripcion="Explora, descubre y administra a tus personajes favoritos en un solo lugar."
      />

      <CustomBreadcrumb
        currentPage="Buscador de heroes"
        breadcrumbs={[
          { label: "Superhéroes", to: "/heroes" }, // Cambia a la ruta correcta de superhéroes
        ]}
      />

      <HeroStats />
      {/* filter and search */}
      <SearchControls />

      {/* heroes */}

      <HeroGrid heroes={heroes} />
    </>
  );
};

export default SearchPage;
