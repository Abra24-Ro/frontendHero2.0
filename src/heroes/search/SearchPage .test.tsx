import { render, screen, waitFor } from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import SearchPage from "./SearchPage";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MemoryRouter } from "react-router";
import { searchHeroesAction } from "../actions/search.heros.action";
import type { Hero } from "../types/hero.interface";

vi.mock("../actions/search.heros.action");

const mockSearchHeroesAction = vi.mocked(searchHeroesAction);

vi.mock("../../components/custom/CustomJumbotron", () => ({
  CustomJumbotron: () => <div data-testid="custom-jumbotron"></div>,
}));

vi.mock("../components/HeroGrid", () => ({
  HeroGrid: ({ heroes }: { heroes: Hero[] }) => (
    <div data-testid="hero-grid">
      {heroes.map((hero) => (
        <div key={hero.id}>{hero.name}</div>
      ))}
    </div>
  ),
}));

vi.mock("./ui/SearchControls", () => ({
  default: () => <div data-testid="search-controls"></div>,
}));

const queryCliente = new QueryClient();

const renderSearchPage = (initialEntries: string[] = ["/"]) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <QueryClientProvider client={queryCliente}>
        <SearchPage />
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe("SearchPage", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should render search page with default values", () => {
    const { container } = renderSearchPage();

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: "",
      strength: "",
    });

    expect(container).toMatchSnapshot();
  });

  test("should call search action with same params", () => {
    const { container } = renderSearchPage(["/search?name=superman"]);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: "superman",
      strength: "",
    });

    expect(container).toMatchSnapshot();
  });

  test("should call search action with strength param", () => {
    const { container } = renderSearchPage(["/search?strength=6"]);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: "",
      strength: "6",
    });

    expect(container).toMatchSnapshot();
  });
  test("should call search action with strength and name parameters", () => {
    const { container } = renderSearchPage(["/search?strength=8&name=batman"]);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: "batman",
      strength: "8",
    });

    expect(container).toMatchSnapshot();
  });

  test("should render HeroGrid with search results", async () => {
    const mockHeroes = [
      {
        id: 1,
        name: "Clark Kent",
      } as unknown as Hero,
      {
        id: 2,
        name: "Bruce Wayne",
      } as unknown as Hero,
    ];
    mockSearchHeroesAction.mockResolvedValue(mockHeroes);
    renderSearchPage();
    await waitFor(() => {
      expect(screen.getByText("Clark Kent")).toBeDefined();
      screen.debug(screen.getByTestId("hero-grid"));
    });
  });
});
