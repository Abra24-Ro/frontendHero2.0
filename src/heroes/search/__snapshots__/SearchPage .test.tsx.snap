// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should call search action with same params 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-4"
    data-slot="breadcrumb"
  >
    <ol
      class="flex-wrap break-words sm:gap-2.5 flex items-center gap-2 text-sm font-medium text-gray-500"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors transition-colors hover:text-green-600"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors cursor-pointer transition-colors cursor-pointer hover:text-green-600 "
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes"
        >
          Superhéroes
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="transition-colors text-black hover:text-green-600"
          data-slot="breadcrumb-link"
        >
          Buscador de heroes
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="text-2xl font-bold text-center"
  >
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strength and name parameters 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-4"
    data-slot="breadcrumb"
  >
    <ol
      class="flex-wrap break-words sm:gap-2.5 flex items-center gap-2 text-sm font-medium text-gray-500"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors transition-colors hover:text-green-600"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors cursor-pointer transition-colors cursor-pointer hover:text-green-600 "
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes"
        >
          Superhéroes
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="transition-colors text-black hover:text-green-600"
          data-slot="breadcrumb-link"
        >
          Buscador de heroes
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="text-2xl font-bold text-center"
  >
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strength param 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-4"
    data-slot="breadcrumb"
  >
    <ol
      class="flex-wrap break-words sm:gap-2.5 flex items-center gap-2 text-sm font-medium text-gray-500"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors transition-colors hover:text-green-600"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors cursor-pointer transition-colors cursor-pointer hover:text-green-600 "
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes"
        >
          Superhéroes
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="transition-colors text-black hover:text-green-600"
          data-slot="breadcrumb-link"
        >
          Buscador de heroes
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="text-2xl font-bold text-center"
  >
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should render search page with default values 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-4"
    data-slot="breadcrumb"
  >
    <ol
      class="flex-wrap break-words sm:gap-2.5 flex items-center gap-2 text-sm font-medium text-gray-500"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors transition-colors hover:text-green-600"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors cursor-pointer transition-colors cursor-pointer hover:text-green-600 "
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes"
        >
          Superhéroes
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash h-4 w-4 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="transition-colors text-black hover:text-green-600"
          data-slot="breadcrumb-link"
        >
          Buscador de heroes
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="text-2xl font-bold text-center"
  >
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;
